File: 02-two-numbers.py
import math

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        root = ListNode()
        current = root
        carry: int = 0

        while(l1 != None || l2 != None || carry != 0):
            l1val = 0
            if (l1 != None):
                l1val = l1.val
                l1 = l1.next

            l2val = 0
            if (l2 != None):
                l2val = l2.val
                l2 = l2.next

            current.val = l1val + l2val + carry
            current.next = ListNode()
            current = current.next

        return root

def main():
    print(listToNum(numToList(342 + 465)))
    print(listToNum(numToList(708)))

if __name__ == "__main__":
    main()

File: 02-two-numbers.cpp
#include <iostream>

//  Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
 public:
  ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    ListNode* root = new ListNode();
    ListNode* current = root;
    int carry = 0;
    while (l1 || l2 || carry) {
      int l1val = l1 ? l1->val : 0;
      int l2val = l2 ? l2->val : 0;

      int val = l1val + l2val + carry;
      carry = 0;
      if (val >= 10) {
        carry = val / 10;
        val %= 10;
      }

      current->next = new ListNode(val);
      current = current->next;

      if (l1) l1 = l1->next;
      if (l2) l2 = l2->next;
    }

    return root->next;
  }
};

int main() {
  auto l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
  auto l2 = new ListNode(5, new ListNode(6, new ListNode(4)));
  Solution solution;

  auto sol = solution.addTwoNumbers(l1, l2);
  std::cout << sol << std::endl;
}